src/epoch_credits/mod.rs

```diff
@@ -1,35 +1,176 @@
 use anyhow::anyhow;
+use prometheus_exporter::prometheus::GaugeVec;
 use solana_client::rpc_client::RpcClient;
-use solana_sdk::clock::Epoch;
-use solana_transaction_status::Rewards;
+use solana_runtime::bank::RewardType;
+use solana_sdk::{
+    clock::Epoch, epoch_info::EpochInfo, native_token::LAMPORTS_PER_SOL, pubkey::Pubkey,
+};
+use solana_stake_program::stake_state::StakeState;
+use solana_transaction_status::{Reward, Rewards};
+use std::{collections::BTreeSet, convert::TryFrom, u64};
 
 pub mod caching;
 pub mod caching_metadata;
 
 const SLOT_OFFSET: u64 = 20;
 
-pub fn get_rewards_for_epoch(epoch: Epoch, client: &RpcClient) -> anyhow::Result<Rewards> {
-    let info = client.get_epoch_info()?;
-
-    // Convert epoch number to slot
-    let start_slot = epoch * info.slots_in_epoch;
-
-    // We cannot use an excessively large range if the epoch just started. There is a chance that
-    // the end slot has not been reached and strange behaviour will occur.
-    // If this is the current epoch and less than `SLOT_OFFSET` slots have elapsed, then do not define an
-    // end_slot for use in the RPC call.
-    let end_slot = if info.epoch == epoch && info.slot_index < SLOT_OFFSET {
-        None
-    } else {
-        Some(start_slot + SLOT_OFFSET)
-    };
-
-    // First block only
-    let block = client
-        .get_blocks(start_slot, end_slot)?
-        .get(0)
-        .cloned()
-        .ok_or_else(|| anyhow!("no blocks found"))?;
-
-    Ok(client.get_block(block)?.rewards)
+struct StakingApy {
+    voter: String,
+    percent: f64,
+}
+
+struct ValidatorReward {
+    voter: String,
+    lamports: u64,
+}
+
+/// The monitor of rewards paid to validators and delegators.
+pub struct RewardsMonitor<'a> {
+    /// Shared Solana RPC client.
+    client: &'a RpcClient,
+    /// Prometheus staking APY gauge.
+    staking_apys: &'a GaugeVec,
+    /// Prometheus cumulative validator rewards gauge.
+    validator_rewards: &'a GaugeVec,
+    /// The last epoch in which rewards were recorded. 0 if uninitialised.
+    ///
+    /// FIXME: cached epoch credit history.
+    last_rewards_epoch: Epoch,
```

Comment:
Note: After the database has been refactored to store staking rewards, this should be replaced by the cache struct.

Reply:
This is the FIXME above. By "epoch credit" I referred to this module `epoch_credits`.

BTW would it make sense to rename the module to `rewards`?

Reply:
Yep, I think that's a better name.